### PR TITLE
Fixed the issue when unauthorized the Pinterest registered app.

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
Fixed the issue when unauthorized the Pinterest registered app.

## Goals
Adding Pinterest Docs
## Approach
N/A
## User stories
N/A
## Release note
N/A
## Documentation
N/A
## Training
N/A

## Certification
N/A
## Marketing
N/A
## Automation tests
N/A
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
N/A
## Related PRs
N/A
## Migrations (if applicable)
N/A
## Test environment
jdk 1.8
ubuntu

## Learning
N/A

### CODE REVIEW

File: component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java [hunk 1]
```diff
@@ -45,351 +45,378 @@
...
-	public boolean canHandle(HttpServletRequest request) {
-		if (log.isDebugEnabled()) {
-			log.debug("Inside PinterestOAuth2Authenticator canHandle method and checking whether the code and state " +
-			          "exist");
-		}
-		return request.getParameter(PinterestAuthenticatorConstants.OAUTH2_GRANT_TYPE_CODE) != null &&
-		       request.getParameter(PinterestAuthenticatorConstants.OAUTH2_PARAM_STATE) != null;
-	}
+    /**
+     * This method is to get Pinterest authorization endpoint.
```
Comment (Biruntha) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java]: Add parameter description also

File: component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java [hunk 2]
```diff
@@ -45,351 +45,378 @@
...
+            String callbackUrl = getCallbackUrl(authenticatorProperties);
+            OAuthAuthzResponse authzResponse = OAuthAuthzResponse.oauthCodeAuthzResponse(request);
+            String code = authzResponse.getCode();
+            OAuthClientRequest accessRequest;
+            try {
+                accessRequest =
+                        OAuthClientRequest.tokenLocation(tokenEndPoint).setGrantType(GrantType.AUTHORIZATION_CODE)
+                                .setClientId(clientId).setClientSecret(clientSecret)
+                                .setRedirectURI(callbackUrl).setCode(code).buildBodyMessage();
+                // create OAuth client that uses custom http client under the hood
```
Comment (Biruntha) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java]: Start with capital letter "C"

File: component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java [hunk 3]
```diff
@@ -45,351 +45,378 @@
...
+                OAuthClient oAuthClient = new OAuthClient(new URLConnectionClient());
+                OAuthClientResponse oAuthResponse;
+                oAuthResponse = oAuthClient.accessToken(accessRequest);
+                String accessToken = oAuthResponse.getParam(PinterestAuthenticatorConstants.ACCESS_TOKEN);
+                if (StringUtils.isNotEmpty(accessToken)) {
+                    Map<ClaimMapping, String> claims = buildClaims(oAuthResponse, authenticatorProperties);
+                    if (claims != null && !claims.isEmpty()) {
+                        //Find the subject from the IDP claim mapping, subject Claim URI.
```
Comment (Biruntha) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java]: put space after //

File: component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java [hunk 4]
```diff
@@ -45,351 +45,378 @@
...
+     * This method is to build the claims for the user info.
+     *
+     * @param token                   token
+     * @param authenticatorProperties authenticatorProperties
+     * @return claims
+     */
+    private Map<ClaimMapping, String> buildClaims(OAuthClientResponse token,
+                                                  Map<String, String> authenticatorProperties)
+            throws ApplicationAuthenticatorException {
+        Map<ClaimMapping, String> claims = new HashMap<>();
```
Comment (Biruntha) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java]: Add one line for better readablility

File: component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java [hunk 5]
```diff
@@ -45,351 +45,378 @@
...
+        String accessToken = token.getParam("access_token");
```
Comment (Biruntha) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java]: Define a constant and use it

File: component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java [hunk 6]
```diff
@@ -45,351 +45,378 @@
...
+    /**
+     * This method is to configure the subject identifier from the claims.
+     *
+     * @param context           AuthenticationContext
+     * @param subjectFromClaims subject identifier claim
+     * @param claims            claims
+     */
+    private void associateSubjectFromClaims(AuthenticationContext context, String subjectFromClaims,
+                                            Map<ClaimMapping, String> claims) {
+        //Use default claim URI on the Authenticator if claim mapping is not defined by the admin
```
Comment (Biruntha) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java]: put space after //

File: component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java [hunk 7]
```diff
@@ -45,351 +45,378 @@
...
-	/**
-	 * This method is to get Pinterest authorization endpoint.
-	 */
-	@Override
-	protected String getAuthorizationServerEndpoint(Map<String, String> authenticatorProperties) {
-		return PinterestAuthenticatorConstants.PINTEREST_OAUTH_ENDPOINT;
-	}
+    /**
+     * This method is to get Pinterest token endpoint.
+     */
```
Comment (Biruntha) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java]: Add parameter description

**Review by keerthu — APPROVED** [review]

**Pull request merged**